### ISSUE TITLE
Fail to deserialize `TimeStamp` .

### ISSUE THREAD

**Issue by weiby3** [mongodb/bson-rust#122]
I want to deserialize str to my struct, but it gets error. The sample code is:
```rust
use serde::{Serialize, Deserialize};
#[derive(Serialize, Deserialize, Debug)]
struct Foo {
    ts: bson::TimeStamp,
}
fn main() -> Result<(), Box<dyn Error>> {
    let original_str = "{\"ts\":1403964864000}";
    let decoded_json = dbg!(serde_json::from_str::<Foo>(original_str)?);
    Ok(())
}
```
Its output is:
```
Error: Error("invalid type: integer `1403964864000`, expected a signed integer", line: 1, column: 19)
```

**Comment by zonyitoo** [comment]
`bson::TimeStamp` is only for converting from/to `Bson`.

**Comment by saghm** [comment]
Hello @weiby3! To clarify further, the `TimeStamp` type is for the the timestamp type as described [here](http://bsonspec.org/spec.html); this is not a type that exists in JSON, so you would need to use extended JSON to deserialize it into Bson. Although we don't currently have a spec-compliant extended JSON implementation in this library, as a workaround you can use a JSON object with the fields `"t"` and `"i"`, which will be converted to a BSON timestamp when the `From<serde_json::Value>` implementation on `Bson` is used:

```rust
use bson::Bson;

fn main() {
    let json = serde_json::json!({
        "t": 12345,
        "i": 54321
    });

    println!("{}", Bson::from(json));
}
```

**Comment by patrickfreed** [comment]
In the 1.0.0 release, a spec compliant extended JSON implementation was completed, so the following should work now (note that `try_from` must be used instead of `from`):
```rust
let json = json!({ "$timestamp": { "t": 12345, "i": 54321 }});
println!("{:?}", bson::Bson::try_from(json).unwrap());
```
I filed RUST-507 to track work for enabling the functionality requested in the original issue (deserializing numbers from JSON or other formats).

**Comment by agolin95** [comment]
Work in RUST-507 is complete! This is available as part of the 1.2.0 release. Please let us know if you encounter any issues!
